### PR TITLE
Redundancy Elimination

### PR DESCRIPTION
**Pull request by Flakebi**
Add redundancy elimination analyses and transformations.
Also add an ant build file to build our project with ant.

### CODE REVIEW

File: analysis/analysis/redundancyelimination/AvailableExpressionAnalysis.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    @Override
+    public Set<Expression> visit(Assignment a, Set<Expression> d) {
+        Set<Expression> newSet = new HashSet<>();
+        if (d != null)
+            newSet.addAll(d);
+        applyTransition(a, newSet);
+
+        Set<Expression> old = dataflowOf(a);
+        if (!lessoreq(newSet, old)) {
+            newSet = lub(newSet, old);
```
Comment (bpoettinger) [inline analysis/analysis/redundancyelimination/AvailableExpressionAnalysis.java]: The `lub` above should not be necessary due to the monotonicity of the edge effects and the lack of joins on edges? Joins only occur on states.

Storing the values on edges is generally not required as we consider values at states and the framework cares about propagation of the value the destination states?

File: analysis/analysis/redundancyelimination/AvailableExpressionAnalysis.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            dataflowOf(a, newSet);
+            return newSet;
+        }
+        return null;
+    }
+
+    @Override
+    public Set<Expression> visit(State s, Set<Expression> d) {
+        if (d != null) {
```
Comment (bpoettinger) [inline analysis/analysis/redundancyelimination/AvailableExpressionAnalysis.java]: I think you are missing joins here. As far as I remember for each change of a value of an incoming edge this method is called once. Without joining the old and the new value you omit the effect of other incoming edges.

Comment (bpoettinger) [reply analysis/analysis/redundancyelimination/AvailableExpressionAnalysis.java]: The other analysis contains the `lub` here.

**Review by bpoettinger — CHANGES_REQUESTED** [review]
I think the avaiable expressions analysis misses a `lub` in the state visitor. The `lub` on the edges should be unnecessary, but should not affect the correctness. The other stuff seems OK.

File: analysis/analysis/redundancyelimination/AvailableExpressionAnalysis.java (hunk 2 above)
Comment (Flakebi) [reply analysis/analysis/redundancyelimination/AvailableExpressionAnalysis.java]: Good catch, thanks!

**Pull request merged**